File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-2.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode mergeTwoLists(ListNode a, ListNode b) {
        var head = new ListNode();
        var tail = head;
        while (a != null || b != null) {
            if (a == null) {
                tail.next = b;
                b = b.next;
            } else if (b == null) {
                tail.next = a;
                a = a.next;
            } else if (a.val > b.val) {
                tail.next = b;
                b = b.next;
            } else {
                tail.next = a;
                a = a.next;
            }
            tail = tail.next;
        }
        return head.next;
    }
}

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:    
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        if not lists or len(lists) == 0:
            return None

        while len(lists) > 1:
            merged = []
            for i in range(0, len(lists), 2):
                a = lists[i]
                b = lists[i + 1] if (i + 1) < len(lists) else None
                merged.append(self.mergeLists(a, b))
            lists = merged
        
        return lists[0]

    def mergeLists(self, a, b):
        dummy = ListNode()
        current = dummy
        while a or b:
            if b == None:
                current.next = ListNode(a.val)
                a = a.next
            elif a == None:
                current.next = ListNode(b.val)
                b = b.next
            elif a.val < b.val:
                current.next = ListNode(a.val)
                a = a.next
            else:
                current.next = ListNode(b.val)
                b = b.next
            current = current.next
        return dummy.next

        